ideas/000-discoverable-trusted-server-nodes/README.md

```diff
@@ -0,0 +1,73 @@
+---
+id: 000-discoverable-trusted-server-nodes
+title: Discoverable And Trusted Server Nodes
+status: Draft
+created: 2018-06-07
+category: core
+lead-contributor: adambabik
+contributors:
+exit-criteria: yes
+success-metrics: yes
+clear-roles: no
+future-iterations: yes
+roles-needed:
+    - Smart Contract Developer
+    - Go Developer
+    - Clojure Developer
+    - UI/UX Designer
+okrs:
+    - "[PO]: Objective: Messaging is reliable"
+    - "[P2]: KR: Cluster can handle 500 concurrent users"
+---
+
+## Summary
+The idea here is to provide a way to make Mail Server nodes discoverable and trusted. This should work seamlessly for the end user.
+
+## Product Overview
+Currently, Mail Servers (and other planned server nodes like Push Notification servers) need to be hardcoded in the app code. It's possible to provide your own Mail Server in the app settings but that's not a scalable solution and it does not solve all issues.
+
+The outcome of this swarm are fully decentralized and discoverable Mail Servers. The Status app will have a mean to verify that a discovered Mail Server is trusted.
+
+In the next iterations, we may work on a way to introduce external providers to submit their own Mail Servers into the cluster and allow end users to accept these resources as trusted.
+
+## Product Description
+The work in this swarm can be divided into three areas:
+- (Infra) making Mail servers discoverable,
+- (Smart Contract) writing a smart contract to keep track of trusted Mail Servers,
+- (Client) mechanism of verifying whether a discovered Mail Server is trusted.
+
+As mentioned in the Product Overview, the smart contract should allow registering custom providers. The client will have an option to decide which providers it trust. Status will be a default provider and by default the app will trust it.
+
+Due to the fact that all Mail Servers are independent and keep the same history of messages, we will be able to remove all hardcoded Mail Server addresses from the client and rely only on discovered Mail Servers. This will allow Status to dynamically scale number of their the Mail Servers depending on the usage.
```

Comment:
I can improve wording here. But overall, all Mail Servers should store the same set of messages. That's the expectation at least. The problem with Whisper is that successful message sent does not mean the message was received and saved by any Mail Server. Now, the question is, can we guarantee that anyhow?

Reply:
@adambabik I think to make any guarantee on consistency of the database in the cluster requires a considerable amount of work, the problem becomes the same as having a distributed database that uses a gossip protocol for data propagation/replication (consul it's an example of this swim/serf, although our case is simpler as append only, but that's just the same as any db oplog, ish).

 Some of the immediate questions we'll have to face: how do you handle new nodes? how do you handle a mailserver being offline for some time? Message loss etc. 

I feel this is a case where we can only provide best-effort, or as @adriacidre suggested plug in an already existing solution that provides a decentralized data store, leaving the mailserver really as a light interface on top of it (there are some other issue to solve, but it seems like they are easier to solve). 

Personally at the moment I think providing best-effort for mailserver consistency is good enough, but worth thinking about how to move forward if we want to.

Reply:
@cammellos 

>  I think to make any guarantee on consistency of the database in the cluster requires a considerable amount of work, the problem becomes the same as having a distributed database that uses a gossip protocol for data propagation/replication

Oh yeah, I haven't been even thinking about this issue. For the time being, Mail Servers should stay as independent storages with their own databases.

I meant the problem intrinsic to Whisper. In a solution with centralized servers, when a message is sent, it's confirmed that it reached the system. In Whisper, it is not a case, hence, we have retries in the client if there is no ack from the recipient (what with public chats though? what if the recipient is offline? We don't get acks from Mail Servers).

> Some of the immediate questions we'll have to face: how do you handle new nodes? how do you handle a mailserver being offline for some time? Message loss etc.

This solution answers the first question. They will be distributed via Discovery V5 and verified that they are trusted using a smart contract.

If a Mail Servers goes down for a few seconds, that's not a big deal because it will receive the messages again. If it's down for more than messages TTL, then the loss of messages is highly likely.

> as @adriacidre suggested plug in an already existing solution that provides a decentralized data store, leaving the mailserver really as a light interface on top of it

That's something that we may explore eventually. Then we can have a front-end of e.g. 3 Mail Servers that run on the same underlying datastore.

> Personally at the moment I think providing best-effort for mailserver consistency is good enough, but worth thinking about how to move forward if we want to.

I agree with that.

Reply:
Small clarification:

> how do you handle new nodes? 
I meant in the scope of database consistency, not discovery, meaning, when I add a new node, how do I make it catch up with the existing nodes, so they all have the same data (i.e adding a new replica)?

In the meantime I am having a look at the smart contracts events you mentioned, just to check what are the options on our side.

Reply:
I see. That is not solved, unfortunately. 

> In the meantime I am having a look at the smart contracts events you mentioned, just to check what are the options on our side.

Thanks!